### DISCUSSION TITLE
Error validating model "Post". Ambiguous relation detected.

### DISCUSSION THREAD

**Discussion by joaocasarin** [prisma/prisma#14326] [Q&A] [answered]
So, I'm trying to create a simple app that has the table `Users` and `Posts`. Basically, any user can create a post, and any user can like any post.
Each post has the `createdAt`, `author`, `authorId` and `likedBy`. Intuitively, what I would add to the `Posts` model would be just `likedBy   User[]`, but because I have already added the relation to `Users` for registering who is the author with `author    User     @relation(fields: [authorId], references: [id])`, prisma raises an `ambiguous relation` problem.

The full prisma models are below.

```prisma
model User {
    id       Int    @id @default(autoincrement())
    name     String
    email    String
    password String
    posts    Post[]
}

model Post {
    id        Int      @id @default(autoincrement())
    text      String
    createdAt DateTime
    author    User     @relation(fields: [authorId], references: [id])
    authorId  Int
    likedBy   User[]
}
```

Honestly, I did not understand why I need to create both `author` and `authorId` in the `Posts` model, since I just need the owner id. But the main question is, after getting the following error:
`Error validating model "Post": Ambiguous relation detected. The fields `author` and `likedBy` in model `Post` both refer to `User`. Please provide different relation names for them by adding `@relation(<name>).`, what would be the correct approach to fix that and get that relation working?

**Comment by TasinIshmam** [comment, marked as answer]
Hi @joaocasarin 

The reason Prisma is raising an issue is that you're trying to add two relations between the same pair of models. This is confusing for Prisma because it does not know which relation field belongs to which model. The way to solve this is to give names to the relations. 

Here is how you can solve the issue: 

```
model User {
  id         Int    @id @default(autoincrement())
  name       String
  email      String
  password   String
  posts      Post[] @relation("WrittenBy")
  likedPosts Post[] @relation("LikedBy")

}

model Post {
  id        Int      @id @default(autoincrement())
  text      String
  createdAt DateTime
  author    User     @relation("WrittenBy", fields: [authorId], references: [id])
  authorId  Int
  likedBy   User[]   @relation("LikedBy")
}

```

More information is available in the [disambiguating relations section](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#disambiguating-relations) of the docs. 

> Honestly, I did not understand why I need to create both author and authorId in the Posts model, since I just need the owner id. 

The `authorId` will hold the foreign key that matches the `User.id` field of the author of a post. This is an actual column in the `Post` table inside your database. The `author` field is a _relation field_ which represents the actual `User` post author record. This is not actually stored in the `Post` table in the underlying database but something Prisma handles itself (by fetching the `User` record where `Post.authorId == User.id`). 

More information is available about this [in the docs](https://www.prisma.io/docs/concepts/components/prisma-schema/relations#relation-fields-1).

**Reply by joaocasarin** [reply]
I'm not sure if I got it correctly. I mean, ok, it will be the FK in the `Post` table, what about the `likedBy` entry in the sabe table, will it be an actual column or just something that Prisma will handle too?

Regarding the two entries in the `User` table:
```prisma
posts      Post[] @relation("CreatedBy")
likedPosts Post[] @relation("LikedBy")
```

Will they become an actual table or just something internal to Prisma? Is there any way to know that?


Anyways, thanks for the reply!

**Reply by TasinIshmam** [reply]
All Prisma model fields where the type is another Prisma model is a relation field. So `Post.likedBy` is also a relation field. They technically are not "real" columns in the database. These relation fields are "handled" by Prisma in different ways, based on the type of the relation. 

For this particular schema: 

1. The `WrittenBy` relation is a one-to-many relation (Each user/author has many posts, and each post has one author). In such cases, you have to mention the foreign key on the "one" side of the relation. For this reason, you have the `Post.authorId` field.  
2. The `likedBy` relation is a many-to-many relation (Each user can like many posts, and each post can be liked by many users). In such cases, you need something called a "pivot table" in the underlying database. This table basically records all the "user-post" combinations for likes. For the schema above, this pivot table is created and handled internally by Prisma. (Note it's possible also to define the pivot table explicitly as a Prisma model). 

To get a better idea, I would suggest reading these articles from the docs:
1.  [Relations](https://www.prisma.io/docs/concepts/components/prisma-schema/relations) 
2. [One-to-many relations](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/one-to-many-relations)
3. [Many-to-many relations](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations)

I hope this clears things up a bit. Let me know if you have any other questions @joaocasarin

**Reply by joaocasarin** [reply]
Hey @TasinIshmam , I really appreciate your effort here helping me understand a bit about prisma. Thank you for both responses, for the suggestion and also for being opened for future questions.

**Reply by TasinIshmam** [reply]
Happy to help 😃

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
